haigha/transports/socket_transport.py

```diff
@@ -29,23 +29,33 @@ def connect(self, (host, port), klass=socket.socket):
         Connect assuming a host and port tuple.
         '''
         self._host = "%s:%s" % (host, port)
-        self._sock = klass()
+        infos = socket.getaddrinfo(host, port, 0, 0, socket.IPPROTO_TCP)
+        if not infos:
+            raise socket.error("getaddrinfo returns an empty list")
+
+        # Sort addresses such that the new IPv6 address is first.
+        for info in reversed(sorted(infos, key=lambda info: len(info[-1]))):
+
+            sockaddr = info[-1]
+            try:
+                # Check if the IP is a valid IPv6 address. Fallback to IPv4 if not.
+                socket.inet_pton(socket.AF_INET6, sockaddr[0])
```

Comment:
Since `getaddrinfo` returns `(family, socktype, proto, canonname, sockaddr)`, what is the purpose of second-guessing the address family here? Per doc,

> In these tuples, family, socktype, proto are all integers and are meant to be passed to the `socket()` function. canonname will be a string representing the canonical name of the host if AI_CANONNAME is part of the flags argument; else canonname will be empty. sockaddr is a tuple describing a socket address, whose format depends on the returned family (a (address, port) 2-tuple for AF_INET, a (address, port, flow info, scope id) 4-tuple for AF_INET6), and is meant to be passed to the `socket.connect()` method.